File: packages/telemetry/src/query/build-promql.ts
import type { GatePolicyInput, GateRunContext } from '@ff-promo/contracts';
import { UnsupportedMetricTypeError } from '../errors/telemetry-adapter-error.js';

export type Cohort = 'treatment' | 'control';

export function escapePromqlLabelValue(value: string): string {
	return value.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
}

function windowSuffix(policy: GatePolicyInput): string {
	const windowSeconds = policy.windowSeconds ?? 300;
	return `[${windowSeconds}s]`;
}

function cohortVariationId(
	runContext: GateRunContext,
	cohort: Cohort,
): string {
	return cohort === 'treatment'
		? runContext.treatmentVariationId
		: runContext.controlVariationId;
}

function labelSelector(
	policy: GatePolicyInput,
	runContext: GateRunContext,
	cohort: Cohort,
	extraLabels = '',
): string {
	const service = escapePromqlLabelValue(policy.serviceName);
	const flagKey = escapePromqlLabelValue(runContext.flagKey);
	const variationId = escapePromqlLabelValue(
		cohortVariationId(runContext, cohort),
	);
	const base = `service="${service}",ld_flag_key="${flagKey}",ld_variation_id="${variationId}",ld_context_kind="user"`;
	if (extraLabels) {
		return `{${base},${extraLabels}}`;
	}
	return `{${base}}`;
}

export function buildErrorRateQuery(
	policy: GatePolicyInput,
	runContext: GateRunContext,
	cohort: Cohort,
): string {
	const labels = labelSelector(policy, runContext, cohort);
	const errorLabels = labelSelector(
		policy,
		runContext,
		cohort,
		'status=~"5.."',
	);
	const window = windowSuffix(policy);
	return `sum(rate(http_requests_total${errorLabels}${window})) / sum(rate(http_requests_total${labels}${window}))`;
}

export function buildLatencyP95Query(
	policy: GatePolicyInput,
	runContext: GateRunContext,
	cohort: Cohort,
): string {
	const labels = labelSelector(policy, runContext, cohort);
	const window = windowSuffix(policy);
	return `histogram_quantile(0.95, sum(rate(http_request_duration_seconds_bucket${labels}${window})) by (le)) * 1000`;
}

export function buildSampleCountQuery(
	policy: GatePolicyInput,
	runContext: GateRunContext,
	cohort: Cohort,
): string {
	const labels = labelSelector(policy, runContext, cohort);
	const window = windowSuffix(policy);
	return `sum(increase(http_requests_total${labels}${window}))`;
}

export function buildMetricQuery(
	metricType: string,
	policy: GatePolicyInput,
	runContext: GateRunContext,
	cohort: Cohort,
): string {
	switch (metricType) {
		case 'error_rate':
			return buildErrorRateQuery(policy, runContext, cohort);
		case 'latency_p95':
			return buildLatencyP95Query(policy, runContext, cohort);
		default:
			throw new UnsupportedMetricTypeError(
				`Unsupported metric type: ${metricType}`,
				metricType,
			);
	}
}

File: kotlin/modules/telemetry/src/main/kotlin/com/ffpromo/telemetry/query/BuildPromql.kt
package com.ffpromo.telemetry.query

import com.ffpromo.contracts.GatePolicyInput
import com.ffpromo.contracts.GateRunContext
import com.ffpromo.contracts.MetricType
import com.ffpromo.telemetry.errors.UnsupportedMetricTypeError

enum class Cohort {
    treatment,
    control,
}

fun escapePromqlLabelValue(value: String): String =
    value.replace("\\", "\\\\").replace("\"", "\\\"")

private fun windowSuffix(policy: GatePolicyInput): String {
    val windowSeconds = policy.windowSeconds ?: 300
    return "[${windowSeconds}s]"
}

private fun cohortVariationId(runContext: GateRunContext, cohort: Cohort): String =
    when (cohort) {
        Cohort.treatment -> runContext.treatmentVariationId
        Cohort.control -> runContext.controlVariationId
    }

private fun labelSelector(
    policy: GatePolicyInput,
    runContext: GateRunContext,
    cohort: Cohort,
    extraLabels: String = "",
): String {
    val service = escapePromqlLabelValue(policy.serviceName)
    val flagKey = escapePromqlLabelValue(runContext.flagKey)
    val variationId = escapePromqlLabelValue(cohortVariationId(runContext, cohort))
    val base =
        """service="$service",ld_flag_key="$flagKey",ld_variation_id="$variationId",ld_context_kind="user""""
    return if (extraLabels.isNotEmpty()) {
        "{$base,$extraLabels}"
    } else {
        "{$base}"
    }
}

fun buildErrorRateQuery(
    policy: GatePolicyInput,
    runContext: GateRunContext,
    cohort: Cohort,
): String {
    val labels = labelSelector(policy, runContext, cohort)
    val errorLabels = labelSelector(policy, runContext, cohort, """status=~"5.."""")
    val window = windowSuffix(policy)
    return "sum(rate(http_requests_total$errorLabels$window)) / sum(rate(http_requests_total$labels$window))"
}

fun buildLatencyP95Query(
    policy: GatePolicyInput,
    runContext: GateRunContext,
    cohort: Cohort,
): String {
    val labels = labelSelector(policy, runContext, cohort)
    val window = windowSuffix(policy)
    return "histogram_quantile(0.95, sum(rate(http_request_duration_seconds_bucket$labels$window)) by (le)) * 1000"
}

fun buildSampleCountQuery(
    policy: GatePolicyInput,
    runContext: GateRunContext,
    cohort: Cohort,
): String {
    val labels = labelSelector(policy, runContext, cohort)
    val window = windowSuffix(policy)
    return "sum(increase(http_requests_total$labels$window))"
}

fun buildMetricQuery(
    metricType: MetricType,
    policy: GatePolicyInput,
    runContext: GateRunContext,
    cohort: Cohort,
): String =
    when (metricType) {
        MetricType.error_rate -> buildErrorRateQuery(policy, runContext, cohort)
        MetricType.latency_p95 -> buildLatencyP95Query(policy, runContext, cohort)
        else -> throw UnsupportedMetricTypeError(
            "Unsupported metric type: $metricType",
            metricType.name,
        )
    }

fun buildMetricQuery(
    metricType: String,
    policy: GatePolicyInput,
    runContext: GateRunContext,
    cohort: Cohort,
): String {
    val resolved = MetricType.entries.find { it.name == metricType }
        ?: throw UnsupportedMetricTypeError(
            "Unsupported metric type: $metricType",
            metricType,
        )
    return buildMetricQuery(resolved, policy, runContext, cohort)
}
